File: Flutter/platinumaps_flutter_sdk/ios/platinumaps_flutter_sdk/Sources/platinumaps_flutter_sdk/PlatinumapsPlatformView.swift
import Flutter
import UIKit

/// Hosts a `PMMapView` inside a Flutter PlatformView and bridges the
/// `PMMapViewDelegate.openLink` callback through a per-instance
/// `FlutterMethodChannel`.
///
/// PlatformView callbacks (`view()`, `init(frame:viewIdentifier:…)`,
/// `applyCreationArguments`) are invoked on the main thread by
/// Flutter; mark the class `@MainActor` so the Swift 6 strict
/// concurrency checker can see that.
@MainActor
final class PlatinumapsPlatformView: NSObject, FlutterPlatformView, PMMapViewDelegate {

    /// URL schemes the plugin will forward to `PMMapView.launchURL`.
    /// Matches the SDK's `browseAllowedSchemes` so the same
    /// allowlist applies to host-supplied launch URLs.
    private static let allowedLaunchUrlSchemes: Set<String> = [
        "http", "https", "tel", "mailto", "sms", "geo",
    ]

    /// `true` when [url]'s scheme is in the launch-URL allowlist.
    /// Exposed at static scope so unit tests can pin the allowlist
    /// behaviour directly.
    internal static func isAllowedLaunchUrlScheme(_ url: URL) -> Bool {
        return Self.allowedLaunchUrlSchemes.contains(
            url.scheme?.lowercased() ?? ""
        )
    }

    private let mapView: PMMapView
    private let methodChannel: FlutterMethodChannel

    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: [String: Any]?,
        messenger: FlutterBinaryMessenger
    ) {
        let mv = PMMapView(frame: frame)
        Self.applyCreationArguments(args, to: mv)

        self.mapView = mv
        self.methodChannel = FlutterMethodChannel(
            name: "\(PlatinumapsFlutterPlugin.viewType)/\(viewId)",
            binaryMessenger: messenger
        )
        super.init()

        // Always claim the delegate. The Dart side decides per-call
        // whether to handle the link or return 'fallback' to delegate
        // back to PMMapView's built-in routing.
        mv.delegate = self

        methodChannel.setMethodCallHandler { [weak self] call, result in
            guard let self else {
                // `result` must be invoked or the Dart Future hangs.
                result(FlutterError(
                    code: "platform_view_disposed",
                    message: "PlatinumapsPlatformView is no longer alive",
                    details: nil
                ))
                return
            }
            self.handle(call, result: result)
        }
    }

    deinit {
        // Sever the delegate link and tear down the method-channel
        // handler before deallocation so in-flight callbacks cannot
        // fire into a deallocated wrapper. UIView deinit runs on the
        // main thread, so assume main-actor isolation rather than hop.
        MainActor.assumeIsolated {
            mapView.delegate = nil
            methodChannel.setMethodCallHandler(nil)
        }
    }

    /// Handles Dart → native invocations coming from
    /// `PlatinumapsMapController`.
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "pushLaunchUrl":
            guard let args = call.arguments as? [String: Any],
                  let urlString = args["url"] as? String,
                  let url = Self.parseLaunchUrl(urlString),
                  Self.isAllowedLaunchUrlScheme(url) else {
                result(FlutterError(
                    code: "invalid_arguments",
                    message: "pushLaunchUrl requires a `url` with an allowlisted scheme",
                    details: nil
                ))
                return
            }
            mapView.pushLaunchURL(url)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    /// Maps the platform-channel creation-args dictionary onto a
    /// `PMMapView`'s public properties. Pulled out of `init` so unit
    /// tests can exercise the field-by-field translation directly,
    /// without standing up a `FlutterMethodChannel`.
    @MainActor
    internal static func applyCreationArguments(
        _ args: [String: Any]?,
        to mapView: PMMapView
    ) {
        mapView.mapSlug = args?["mapSlug"] as? String

        // Match the wire format the native Android SDK uses: it appends
        // `beaconminsample` / `beaconmaxhistory` / `memo` URL query
        // parameters during `openPlatinumaps`. iOS PMMapView has no
        // public API for these optional fields, so fold them into
        // `mapQuery`. Caller-supplied `queryParams` are merged last so
        // a clashing key (e.g. `memo`) keeps the public-API value.
        var mapQuery: [String: String] = [:]
        if let beacon = args?["beacon"] as? [String: Any] {
            if let uuid = beacon["uuid"] as? String {
                mapView.beaconUuid = uuid
            }
            if let minSample = beacon["minSample"] as? Int {
                mapQuery["beaconminsample"] = String(minSample)
            }
            if let maxHistory = beacon["maxHistory"] as? Int {
                mapQuery["beaconmaxhistory"] = String(maxHistory)
            }
            if let memo = beacon["memo"] as? String {
                mapQuery["memo"] = memo
            }
        }
        if let queryParams = args?["queryParams"] as? [String: String] {
            for (key, value) in queryParams {
                mapQuery[key] = value
            }
        }
        mapView.mapQuery = mapQuery

        if let localeCode = args?["locale"] as? String,
           let locale = PMLocale(rawValue: localeCode) {
            mapView.mapLocale = locale
        }
        if let appStoreId = args?["appStoreId"] as? String {
            mapView.appStoreId = appStoreId
        }
        if let userId = args?["userId"] as? String {
            mapView.userId = userId
        }
        if let secretKey = args?["secretKey"] as? String {
            mapView.secretKey = secretKey
        }
        if let offsetBottom = args?["offsetBottom"] as? Int {
            mapView.offsetBottom = offsetBottom
        }
        if let launchUrlString = args?["launchUrl"] as? String,
           let launchUrl = Self.parseLaunchUrl(launchUrlString),
           Self.allowedLaunchUrlSchemes.contains(launchUrl.scheme?.lowercased() ?? "") {
            // The web layer eventually echoes `launchUrl` back through
            // its own command flow, so without an allowlist a Flutter
            // host could stage a `javascript:` or `data:`-style URL
            // via the Dart API and reach the inside of the
            // PlatformView's WebView. Mirror the SDK's
            // `browseAllowedSchemes` allowlist here so the same
            // defence-in-depth applies to host-supplied launch URLs.
            mapView.launchURL = launchUrl
        }
    }

    func view() -> UIView {
        return mapView
    }

    /// Parse a `launchUrl` string into a `URL`.
    ///
    /// On iOS 17+ the default `URL(string:)` silently percent-encodes
    /// invalid characters, which can turn a malformed Dart-side string
    /// into a "valid" URL. The `encodingInvalidCharacters: false`
    /// overload (also iOS 17+) refuses the input instead — which is
    /// what we want, since the Dart side already typed the value as
    /// `Uri`. On iOS 16 we fall back to the older, lenient behaviour.
    private static func parseLaunchUrl(_ string: String) -> URL? {
        if #available(iOS 17.0, *) {
            return URL(string: string, encodingInvalidCharacters: false)
        }
        return URL(string: string)
    }

    // MARK: - PMMapViewDelegate

    func openLink(_ url: URL, sharedCookie: Bool) {
        openLink(url, sharedCookie: sharedCookie, openInExternalApp: false)
    }

    func openLink(_ url: URL, sharedCookie: Bool, openInExternalApp: Bool) {
        methodChannel.invokeMethod(
            "onOpenLink",
            arguments: [
                "url": url.absoluteString,
                "sharedCookie": sharedCookie,
            ]
        ) { [weak self] reply in
            guard let self else { return }
            if (reply as? String) == "fallback" {
                self.mapView.openLinkUsingDefault(
                    url,
                    sharedCookie: sharedCookie,
                    openInExternalApp: openInExternalApp,
                )
            }
        }
    }
}

File: Flutter/platinumaps_flutter_sdk/android/src/main/kotlin/jp/co/boldright/platinumaps/flutter/PlatinumapsPlatformView.kt
package jp.co.boldright.platinumaps.flutter

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.view.View
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import jp.co.boldright.platinumaps.sdk.PmMapBeaconOptions
import jp.co.boldright.platinumaps.sdk.PmMapOptions
import jp.co.boldright.platinumaps.sdk.PmWebView

/**
 * Hosts a [PmWebView] inside a Flutter PlatformView and bridges the
 * `onOpenLink` callback through a per-instance [MethodChannel].
 */
internal class PlatinumapsPlatformView(
    context: Context,
    messenger: BinaryMessenger,
    private val viewId: Int,
    args: Map<String, Any?>?,
    private val onDispose: (Int) -> Unit,
) : PlatformView, PmWebView.OnOpenLinkListener {

    private val webView: PmWebView = PmWebView(context)
    private val methodChannel: MethodChannel = MethodChannel(
        messenger,
        "${PlatinumapsFlutterPlugin.VIEW_TYPE}/$viewId",
    )

    // `dispose()` and the host Lifecycle's onDestroy both reach the
    // native SDK's `activityDestroy()` (CLAUDE.md: "call exactly once
    // per WebView instance"). Either may fire first, so this flag
    // makes the second call a no-op.
    private var hasDestroyedWebView = false

    init {
        // Always claim the listener. The Dart side decides per-call
        // whether to handle the link or return 'fallback' to delegate
        // back to PmWebView's built-in routing.
        webView.onOpenLinkListener = this
        methodChannel.setMethodCallHandler { call, result -> handle(call, result) }

        val offsetBottom = (args?.get("offsetBottom") as? Number)?.toInt() ?: 0
        val (safeAreaTop, safeAreaBottom) = resolveSafeAreaInsets(
            context = context,
            zeroBottom = offsetBottom > 0,
        )
        webView.openPlatinumaps(
            buildMapOptions(
                args = args,
                safeAreaTop = safeAreaTop,
                safeAreaBottom = safeAreaBottom,
            ),
        )

        // The native Android SDK's `PmMapOptions` has no `launchUrl`
        // field today, so we can't fold this into `buildMapOptions`.
        // Push it through `pushLaunchURL` instead — the WebView is
        // still booting, so `PmWebView` stashes the URL and replays
        // it when `web.ready` arrives, matching the iOS path.
        val launchUrl = args?.get("launchUrl") as? String
        if (launchUrl != null) {
            val uri = Uri.parse(launchUrl)
            if (isAllowedLaunchUrlScheme(uri.scheme)) {
                webView.pushLaunchURL(uri)
            }
        }
    }

    /** Dart → native invocations coming from `PlatinumapsMapController`. */
    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "pushLaunchUrl" -> {
                val urlString = call.argument<String>("url")
                if (urlString == null) {
                    result.error("invalid_arguments", "pushLaunchUrl requires a `url`", null)
                    return
                }
                val uri = Uri.parse(urlString)
                if (!isAllowedLaunchUrlScheme(uri.scheme)) {
                    result.error(
                        "invalid_arguments",
                        "pushLaunchUrl requires a `url` with an allowlisted scheme",
                        null,
                    )
                    return
                }
                webView.pushLaunchURL(uri)
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    companion object {
        /**
         * URL schemes the plugin will forward to
         * `PmWebView.pushLaunchURL`. Matches the native SDK's
         * `browseAllowedSchemes` and the iOS plugin glue's allowlist.
         */
        private val allowedLaunchUrlSchemes: Set<String> = setOf(
            "http", "https", "tel", "mailto", "sms", "geo",
        )

        /**
         * `true` when [scheme] is in the launch-URL allowlist. Pulled
         * out of `handle()` so unit tests can exercise the allowlist
         * without standing up a real WebView. Takes a `String?`
         * (rather than `Uri`) so the test source set can call it
         * without depending on Robolectric for `Uri.parse`.
         */
        internal fun isAllowedLaunchUrlScheme(scheme: String?): Boolean {
            if (scheme == null) return false
            return scheme.lowercase() in allowedLaunchUrlSchemes
        }

        /**
         * Resolves the system-bar + display-cutout insets from the host
         * Activity's decor view, so the WebView can lay out under
         * status bar / nav bar / notch the same way the iOS path does.
         * `zeroBottom` reflects the Dart-side `offsetBottom` flag: when
         * the host already draws a bottom inset (e.g. a tab bar) the
         * map should ignore the system one.
         */
        internal fun resolveSafeAreaInsets(
            context: Context,
            zeroBottom: Boolean,
        ): Pair<Int, Int> {
            val activity = context as? Activity ?: return Pair(0, 0)
            val rootInsets = ViewCompat.getRootWindowInsets(activity.window.decorView)
                ?: return Pair(0, 0)
            val systemInsets = rootInsets.getInsets(
                WindowInsetsCompat.Type.systemBars()
                    or WindowInsetsCompat.Type.displayCutout(),
            )
            return Pair(systemInsets.top, if (zeroBottom) 0 else systemInsets.bottom)
        }

        /**
         * Translates the creation arguments the Dart side sends through
         * the platform channel into a [PmMapOptions] the native SDK
         * understands. Exposed at companion scope (rather than buried in
         * the `init` block) so unit tests can pin its behaviour without
         * standing up a real [PmWebView].
         *
         * `locale` is folded into the `culture` query parameter so the
         * Dart API reaches the Android web layer through the same path
         * the web app already understands. The native Android SDK has
         * no first-class `mapLocale` field today (CLAUDE.md notes that
         * Android derives culture from Accept-Language), but the web
         * app honours an explicit `culture=` override regardless of
         * platform.
         *
         * `launchUrl`, `appStoreId`, and `offsetBottom` (the flag-only
         * value) are consumed by the init block, not here. Safe-area
         * insets are resolved by the caller and passed in.
         */
        internal fun buildMapOptions(
            args: Map<String, Any?>?,
            safeAreaTop: Int = 0,
            safeAreaBottom: Int = 0,
        ): PmMapOptions {
            val mapSlug = (args?.get("mapSlug") as? String).orEmpty()
            if (mapSlug.isEmpty()) {
                Log.w(
                    "PlatinumapsFlutter",
                    "mapSlug is empty; the WebView will load /maps/ and 404.",
                )
            }
            // The Dart side declares `Map<String, String>?` for
            // queryParams, but the platform channel runtime erases the
            // generics: what arrives at the JVM is a `Map<*, *>` whose
            // values come back as `Any?`. Filter explicitly so a stray
            // non-String value cannot crash later inside
            // `appendQueryParameter`.
            val rawQueryParams: Map<String, String>? =
                (args?.get("queryParams") as? Map<*, *>)?.entries
                    ?.mapNotNull { (rawKey, rawValue) ->
                        val key = rawKey as? String ?: return@mapNotNull null
                        val value = rawValue as? String ?: return@mapNotNull null
                        key to value
                    }
                    ?.toMap()
            val locale = args?.get("locale") as? String
            val queryParams = when {
                locale == null -> rawQueryParams
                rawQueryParams == null -> mapOf("culture" to locale)
                else -> rawQueryParams + ("culture" to locale)
            }

            val beaconMap = (args?.get("beacon") as? Map<*, *>)?.let { raw ->
                raw.entries
                    .mapNotNull { (rawKey, rawValue) ->
                        val key = rawKey as? String ?: return@mapNotNull null
                        key to rawValue
                    }
                    .toMap()
            }
            val beaconOptions = beaconMap?.let {
                PmMapBeaconOptions(
                    uuid = (it["uuid"] as? String).orEmpty(),
                    minSample = (it["minSample"] as? Number)?.toInt(),
                    maxHistory = (it["maxHistory"] as? Number)?.toInt(),
                    memo = it["memo"] as? String,
                )
            }

            return PmMapOptions(
                mapPath = mapSlug,
                queryParams = queryParams,
                safeAreaTop = safeAreaTop,
                safeAreaBottom = safeAreaBottom,
                beacon = beaconOptions,
                userId = args?.get("userId") as? String,
                secretKey = args?.get("secretKey") as? String,
            )
        }
    }

    override fun getView(): View = webView

    override fun dispose() {
        methodChannel.setMethodCallHandler(null)
        webView.onOpenLinkListener = null
        destroyWebViewOnce()
        onDispose(viewId)
    }

    /**
     * Forwards Activity `onResume` from the plugin's lifecycle
     * observer. Resume is driven by the host Activity going to the
     * foreground, not by the PlatformView attaching to the Flutter
     * view hierarchy — those events do not coincide.
     */
    fun activityResume() {
        webView.activityResume()
    }

    /**
     * Forwards Activity `onPause` from the plugin's lifecycle
     * observer. Without this, GPS and BLE scanning continue while
     * the app sits in the background.
     */
    fun activityPause() {
        webView.activityPause()
    }

    /**
     * Forwards Activity `onDestroy`. `dispose()` already calls
     * `activityDestroy()` on its own, but the host Activity can go
     * away while the PlatformView is still attached, in which case
     * the lifecycle event arrives first. The guard inside
     * [destroyWebViewOnce] keeps the second call a no-op.
     */
    fun activityDestroy() {
        destroyWebViewOnce()
    }

    private fun destroyWebViewOnce() {
        if (hasDestroyedWebView) return
        hasDestroyedWebView = true
        webView.activityDestroy()
    }

    override fun onOpenLink(url: Uri, sharedCookie: Boolean) {
        onOpenLink(url, sharedCookie, openInExternalApp = false)
    }

    override fun onOpenLink(url: Uri, sharedCookie: Boolean, openInExternalApp: Boolean) {
        methodChannel.invokeMethod(
            "onOpenLink",
            mapOf(
                "url" to url.toString(),
                "sharedCookie" to sharedCookie,
            ),
            object : MethodChannel.Result {
                override fun success(result: Any?) {
                    if (result == "fallback") {
                        webView.openLinkUsingDefault(url, sharedCookie, openInExternalApp)
                    }
                }
                override fun error(code: String, message: String?, details: Any?) {}
                override fun notImplemented() {}
            },
        )
    }

    fun handlePermissionResult(requestCode: Int, grantResults: IntArray) {
        webView.handlePermissionResult(requestCode, grantResults)
    }

    fun handleActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        webView.handleFileChooserResult(requestCode, resultCode, data)
    }
}
